File: hw/maisincho/u3/HW21TemplateMethod/TemplateMeThod/src/main/java/ec/edu/espe/template/model/CaffeineBeverage.java
package ec.edu.espe.template.model;

/**
 *
 * @author Isaac Maisincho Crafters_Market DCCO ESPE
 */
public abstract class CaffeineBeverage {

    public void prepareRecipe() {//Templete Method
        boilWater();
        brew();
        pourInCup();
        if (wantsCondiments()) {
            addCondiments();
        }
    }

    void boilWater() {
        System.out.println("Boling water");
    }

    abstract void brew();

    void pourInCup() {
        System.out.println("Pouring into cup");
    }

    abstract void addCondiments();

    boolean wantsCondiments() {
        return true;
    }

}

File: hw/bonilla/U3/HW21TemplateMethod/TemplateMethodjs/model/CaffeineBeverage.js
class CaffeineBeverage {
    async prepareRecipe() {
        this.boilWater();
        await this.brew();
        this.pourInCup();
        if (await this.wantsCondiments()) {
            this.addCondiments();
        }
    }

    boilWater() {
        console.log("Boiling water");
    }

    // métodos abstractos - deben implementar las subclases
    brew() {
        throw new Error("brew() must be implemented");
    }

    pourInCup() {
        console.log("Pouring into cup");
    }

    addCondiments() {
        throw new Error("addCondiments() must be implemented");
    }

    wantsCondiments() {
        return true;
    }
}

module.exports = CaffeineBeverage;
